File: ios/WealthBoard/Data/DiskCache.swift
import Foundation

/// A small on-disk cache for things that are expensive to fetch and cheap to
/// keep: news headlines, scraped dividend calendars.
///
/// ## Why not the portfolio document
///
/// `PortfolioStore` holds what the user typed — accounts, holdings,
/// transactions, dividends. It is a few kilobytes, it is written atomically,
/// and it is re-saved on every edit. Putting a hundred kilobytes of news
/// headlines in it would mean rewriting all of that every time someone changed
/// a unit count, and would mix data the user would be upset to lose with data
/// that can be thrown away at any moment.
///
/// So each cache gets its own file, in the caches directory rather than
/// application support. That is the honest location: iOS is free to delete it
/// under storage pressure, which is exactly the right outcome for something
/// that can be re-fetched. Nothing here is load-bearing.
///
/// ## Staleness is the caller's decision
///
/// `load` hands back the value AND its age, rather than deciding for itself
/// whether the value is too old. A cold launch wants to paint whatever it has,
/// however old; a pull-to-refresh wants to skip the cache entirely; a
/// background refresh wants to know if it should bother. One store, three
/// policies, so the age travels with the value and the caller picks.
actor DiskCache {

    /// One cached value and when it was written.
    struct Entry<Value: Codable>: Codable {
        let value: Value
        let storedAt: Date

        var age: TimeInterval { Date().timeIntervalSince(storedAt) }

        /// True when the entry is older than `ttl`. A caller that does not
        /// care about age simply ignores this.
        func isStale(after ttl: TimeInterval) -> Bool { age > ttl }
    }

    static let shared = DiskCache()

    /// Named caches, so callers refer to a file by a value rather than a
    /// string literal in three places.
    enum Key: String {
        case marketNews = "market-news"
        case dividendNews = "dividend-news"
        case dividendCalendars = "dividend-calendars"
    }

    private var directory: URL {
        let base = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first
            ?? FileManager.default.temporaryDirectory
        let dir = base.appendingPathComponent("WealthBoardCache", isDirectory: true)
        if !FileManager.default.fileExists(atPath: dir.path) {
            try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        }
        return dir
    }

    private func url(for key: Key) -> URL {
        directory.appendingPathComponent("\(key.rawValue).json")
    }

    private var encoder: JSONEncoder {
        let e = JSONEncoder()
        e.dateEncodingStrategy = .millisecondsSince1970
        return e
    }

    private var decoder: JSONDecoder {
        let d = JSONDecoder()
        d.dateDecodingStrategy = .millisecondsSince1970
        return d
    }

    // MARK: - Reading and writing

    /// The cached value and its age, or nil when there is nothing usable.
    ///
    /// A decode failure is treated as a miss rather than an error: the shape on
    /// disk was written by an older build, and a cache that cannot be read is
    /// simply a cache that is empty.
    func load<Value: Codable>(_ key: Key, as type: Value.Type) -> Entry<Value>? {
        guard let data = try? Data(contentsOf: url(for: key)) else { return nil }
        return try? decoder.decode(Entry<Value>.self, from: data)
    }

    /// Writes atomically, so a crash mid-save leaves the previous good file
    /// rather than a truncated one.
    func save<Value: Codable>(_ key: Key, _ value: Value) {
        guard let data = try? encoder.encode(Entry(value: value, storedAt: Date())) else { return }
        try? data.write(to: url(for: key), options: .atomic)
    }

    func clear(_ key: Key) {
        try? FileManager.default.removeItem(at: url(for: key))
    }

    /// Everything, for a "clear cached data" action or a sign-out.
    func clearAll() {
        for key in [Key.marketNews, .dividendNews, .dividendCalendars] { clear(key) }
    }
}

File: android/app/src/main/java/ca/tristan/portfolio/data/DiskCache.kt
package ca.tristan.portfolio.data

import ca.tristan.portfolio.net.DividendHistoryOrgEntry
import ca.tristan.portfolio.net.NewsCategory
import ca.tristan.portfolio.net.NewsItem
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

/**
 * An on-disk cache for things that are expensive to fetch and cheap to keep:
 * news headlines and scraped dividend calendars.
 *
 * ## Why files rather than Room
 *
 * Room holds what the user typed, and adding tables to it means a schema
 * migration — real risk, for data that can be thrown away at any moment. These
 * caches live as JSON files in `cacheDir`, which is the honest location:
 * Android is free to delete that directory under storage pressure, which is
 * exactly the right outcome for something re-fetchable. Nothing here is
 * load-bearing; every read can return null and every caller copes.
 *
 * ## Why hand-rolled JSON
 *
 * The project has no Gson, Moshi or kotlinx-serialization, and adding one to
 * persist two small shapes would be a dependency for its own sake. `org.json`
 * is already used throughout the networking layer, so this matches how the
 * rest of the code reads and writes JSON.
 *
 * Attached once at startup, the same way [FxRates] is, so the object itself
 * needs no Context.
 */
object DiskCache {

    private const val MARKET_NEWS = "market-news.json"
    private const val DIVIDEND_NEWS = "dividend-news.json"
    private const val DIVIDEND_CALENDARS = "dividend-calendars.json"

    @Volatile
    private var dir: File? = null

    /** Call once at startup with `application.cacheDir`. */
    fun attach(cacheDir: File) {
        val target = File(cacheDir, "wealthboard")
        if (!target.exists()) target.mkdirs()
        dir = target
    }

    private fun file(name: String): File? = dir?.let { File(it, name) }

    /**
     * Reads a cache file and returns its payload with the age of the write.
     *
     * Age travels with the value rather than being checked here, because the
     * three callers want three different policies: a cold launch paints
     * whatever it has however old, a pull-to-refresh skips the cache entirely,
     * and a background refresh wants to know whether to bother.
     */
    private fun read(name: String): Pair<JSONObject, Long>? {
        val f = file(name) ?: return null
        if (!f.exists()) return null
        return try {
            val root = JSONObject(f.readText())
            val storedAt = root.optLong("storedAt", 0L)
            if (storedAt <= 0L) null else root to (System.currentTimeMillis() - storedAt)
        } catch (_: Exception) {
            // A file written by an older build that no longer parses is a miss,
            // not an error. A cache that cannot be read is a cache that is empty.
            null
        }
    }

    private fun write(name: String, body: JSONObject) {
        val f = file(name) ?: return
        try {
            body.put("storedAt", System.currentTimeMillis())
            // Via a temp file, so a crash mid-write leaves the previous good
            // copy rather than a truncated one.
            val tmp = File(f.parentFile, "${f.name}.tmp")
            tmp.writeText(body.toString())
            if (!tmp.renameTo(f)) {
                f.writeText(body.toString())
                tmp.delete()
            }
        } catch (_: Exception) {
            // Best effort by design.
        }
    }

    // ── News ─────────────────────────────────────────────────────────────────

    private fun newsFile(dividend: Boolean) = if (dividend) DIVIDEND_NEWS else MARKET_NEWS

    /**
     * The last headlines fetched, or null when there are none or they are old
     * enough that showing them would mislead.
     *
     * A day. Past that the feed is history rather than news, and an empty tab
     * that fills in beats yesterday's headlines presented as today's.
     */
    fun loadNews(dividend: Boolean, maxAgeMs: Long = 24L * 60 * 60 * 1000): List<NewsItem>? {
        val (root, age) = read(newsFile(dividend)) ?: return null
        if (age > maxAgeMs) return null
        val arr = root.optJSONArray("items") ?: return null
        val out = ArrayList<NewsItem>(arr.length())
        for (i in 0 until arr.length()) {
            val o = arr.optJSONObject(i) ?: continue
            val title = o.optString("title").takeIf { it.isNotBlank() } ?: continue
            val link = o.optString("linkUrl").takeIf { it.isNotBlank() } ?: continue
            out.add(
                NewsItem(
                    title = title,
                    publisher = o.optString("publisher").ifBlank { "Market news" },
                    linkUrl = link,
                    publishedAt = o.optLong("publishedAt", 0L),
                    imageUrl = o.optString("imageUrl").takeIf { it.isNotBlank() },
                    summary = o.optString("summary").takeIf { it.isNotBlank() },
                    category = runCatching {
                        NewsCategory.valueOf(o.optString("category", NewsCategory.MARKETS.name))
                    }.getOrDefault(NewsCategory.MARKETS)
                )
            )
        }
        return out.ifEmpty { null }
    }

    /** Bounded at 60 — the feeds return about that and nobody scrolls further. */
    fun saveNews(dividend: Boolean, items: List<NewsItem>) {
        if (items.isEmpty()) return
        val arr = JSONArray()
        items.take(60).forEach { item ->
            arr.put(
                JSONObject().apply {
                    put("title", item.title)
                    put("publisher", item.publisher)
                    put("linkUrl", item.linkUrl)
                    put("publishedAt", item.publishedAt)
                    item.imageUrl?.let { put("imageUrl", it) }
                    item.summary?.let { put("summary", it) }
                    put("category", item.category.name)
                }
            )
        }
        write(newsFile(dividend), JSONObject().put("items", arr))
    }

    // ── Dividend calendars ───────────────────────────────────────────────────

    /**
     * Every cached payout calendar, keyed by uppercased ticker, with the age of
     * the file.
     *
     * Scraping these is the slowest thing the app does — up to four candidate
     * URLs per ticker — and a fund declares a distribution a few times a year,
     * so re-running it on every launch was the worst effort-to-value ratio in
     * the codebase.
     */
    fun loadCalendars(): Pair<Map<String, List<DividendHistoryOrgEntry>>, Long>? {
        val (root, age) = read(DIVIDEND_CALENDARS) ?: return null
        val byTicker = root.optJSONObject("tickers") ?: return null
        val out = HashMap<String, List<DividendHistoryOrgEntry>>()
        for (ticker in byTicker.keys()) {
            val arr = byTicker.optJSONArray(ticker) ?: continue
            val rows = ArrayList<DividendHistoryOrgEntry>(arr.length())
            for (i in 0 until arr.length()) {
                val o = arr.optJSONObject(i) ?: continue
                val ex = o.optLong("exDateMs", 0L)
                val amount = o.optDouble("amountPerShare", Double.NaN)
                if (ex <= 0L || amount.isNaN()) continue
                rows.add(
                    DividendHistoryOrgEntry(
                        exDateMs = ex,
                        payDateMs = o.optLong("payDateMs", 0L).takeIf { it > 0L },
                        amountPerShare = amount,
                        isEstimated = o.optBoolean("isEstimated", false)
                    )
                )
            }
            if (rows.isNotEmpty()) out[ticker.uppercase()] = rows
        }
        return if (out.isEmpty()) null else out to age
    }

    fun saveCalendars(calendars: Map<String, List<DividendHistoryOrgEntry>>) {
        if (calendars.isEmpty()) return
        val byTicker = JSONObject()
        calendars.forEach { (ticker, rows) ->
            val arr = JSONArray()
            rows.forEach { row ->
                arr.put(
                    JSONObject().apply {
                        put("exDateMs", row.exDateMs)
                        row.payDateMs?.let { put("payDateMs", it) }
                        put("amountPerShare", row.amountPerShare)
                        put("isEstimated", row.isEstimated)
                    }
                )
            }
            byTicker.put(ticker.uppercase(), arr)
        }
        write(DIVIDEND_CALENDARS, JSONObject().put("tickers", byTicker))
    }

    /** For a "clear cached data" action, or a sign-out. */
    fun clearAll() {
        listOf(MARKET_NEWS, DIVIDEND_NEWS, DIVIDEND_CALENDARS).forEach { file(it)?.delete() }
    }
}
